compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -622,27 +634,33 @@ object Semantic {
         (errors, allArgsPromote)
 
       // fast track if the current object is already initialized
-      if promoted.isCurrentObjectPromoted then Result(Hot, Nil)
-      else if isAlwaysSafe(meth) then Result(Hot, Nil)
-      else if meth eq defn.Any_asInstanceOf then Result(value, Nil)
+      if promoted.isCurrentObjectPromoted then Hot
+      else if isAlwaysSafe(meth) then Hot
+      else if meth eq defn.Any_asInstanceOf then value
       else value match {
         case Hot  =>
           if isSyntheticApply(meth) then
             val klass = meth.owner.companionClass.asClass
-            instantiate(klass, klass.primaryConstructor, args, source)
+            instantiate(klass, klass.primaryConstructor, args)
           else
             if receiver.typeSymbol.isStaticOwner then
               val (errors, allArgsPromote) = checkArgsWithParametricity()
-              if allArgsPromote || errors.nonEmpty then
-                Result(Hot, errors)
+              if allArgsPromote then
+                Hot: Value
+              else if errors.nonEmpty then
+                for error <- errors do reporter.report(error)
+                Hot: Value
```

Comment:
Why is the ascription needed?

Reply:
This is due to a TreeChecker failure:

https://github.com/lampepfl/dotty/pull/15263#issuecomment-1134028074

I have no idea why the subtyping failed --- it shouldn't be.